relayer/validate.go

```diff
@@ -0,0 +1,127 @@
+package relayer
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+
+	"github.com/athanorlabs/go-relayer/impls/gsnforwarder"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethclient"
+
+	"github.com/athanorlabs/atomic-swap/coins"
+	"github.com/athanorlabs/atomic-swap/common/types"
+	contracts "github.com/athanorlabs/atomic-swap/ethereum"
+	"github.com/athanorlabs/atomic-swap/net/message"
+)
+
+func validateClaimRequest(
+	ctx context.Context,
+	request *message.RelayClaimRequest,
+	ec *ethclient.Client,
+	expectedForwarderAddress ethcommon.Address,
+) error {
+	err := validateClaimValues(ctx, request, ec, expectedForwarderAddress, MinRelayerFeeWei)
+	if err != nil {
+		return err
+	}
+
+	return validateClaimSignature(ctx, ec, expectedForwarderAddress, request)
+}
+
+// validateClaimValues validates the non-signature aspects of the claim request:
+//  1. the claim request swap factory contract is byte compatible with ours
+//  2. the forwarder in the claim request swap factory contract has an identical
+//     address with our forwarder
+//  3. the relayer fee is equal to or greater than the passed minFee
+//  4. the swap is for ETH and not an ERC20 token
+//  5. the swap value is strictly greater than the relayer fee
+//  6. TODO: Validate that the swap exists and is in a claimable state?
+func validateClaimValues(
+	ctx context.Context,
+	req *message.RelayClaimRequest,
+	ec *ethclient.Client,
+	expectedForwarderAddress ethcommon.Address,
+	minFee *big.Int,
+) error {
+	// Validate that the deployed SwapFactory contract has the same bytecode as
+	// the one we use. There is a good chance that it has the same exact address
+	// as ours, but we check for binary compatibility regardless.
+	requestedForwarderAddr, err := contracts.CheckSwapFactoryContractCode(ctx, ec, req.SFContractAddress)
+	if err != nil {
+		return err
+	}
+
+	// The forwarder used must have the same exact address as ours, so we don't
+	// need to check the forwarder contract bytecode.
+	if requestedForwarderAddr != expectedForwarderAddress {
+		return fmt.Errorf("claim request had expected forwarder address: got %s, expected %s",
```

Comment:
```suggestion
		return fmt.Errorf("claim request had unexpected forwarder address: got %s, expected %s",
```

Reply:
I would argue that using forwarder that aren't the same address as ours is fine, as long as the bytecode matches

Reply:
That seems reasonable. Requiring the exact same forwarder address and always verifying the swap factory contract, even if it has the same address as our own, was what the existing code was doing.

Reply:
The code is updated and no longer requires that the forwarder address be exactly the same as ours. `CheckSwapFactoryContractCode` calls `CheckForwarderContractCode`, so the forwarder doesn't need to have its byte code checked separate from the swap factory contract.